### PR TITLE
Fix TableViewImpl not retrying partition update on exceptions

### PR DESCRIPTION
**Pull request by ZiyaoWei**
### Motivation

TableViewImpl doesn't keep scheduling partition update checks if one check fails exceptionally, so if a call to `start()` fails, it will stop keeping track of future partition updates. This fixes that by manually scheduling a check on exception.

### Modifications

* Added logic to schedule a partition update check on exceptions
* Added a regression test

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

`mvn clean -pl pulsar-client install -Dmaven.test.skip && mvn clean -pl pulsar-broker test -Dtest=TableViewTest#testTableViewUpdatePartitionsKeepTryingOnExceptions`

### Documentation

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  Small fix
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Comment by gaozhangmin** [comment]
Please avoid the unnecessary indent changes @ZiyaoWei

**Review by gaozhangmin — DISMISSED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java [hunk 1]
```diff
@@ -127,12 +128,12 @@ public boolean containsKey(String key) {
 
     @Override
     public T get(String key) {
-       return data.get(key);
+        return data.get(key);
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:131]: Please undo formattting

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java [hunk 2]
```diff
@@ -104,9 +104,10 @@ private void checkForPartitionsChanges(Timeout timeout) {
         }
 
         start().whenComplete((tw, ex) -> {
-           if (ex != null) {
-               log.warn("Failed to check for changes in number of partitions");
-           }
+            if (ex != null) {
+                log.warn("Failed to check for changes in number of partitions");
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:108]: Please report the exception

**Review by eolivelli — CHANGES_REQUESTED** [review]
Thanks for your contribution 

Can you please share the stacktrace of the error?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java (hunk 1 above)
Comment (ZiyaoWei) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java]: Done, thanks @eolivelli and @gaozhangmin! My IntelliJ did the reformatting automatically - is there a code formatter that this repository uses?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java (hunk 2 above)
Comment (ZiyaoWei) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java]: Not sure how to do that - do you mean via Prometheus? Sorry, new to Pulsar.

**Comment by ZiyaoWei** [comment]
@eolivelli, @gaozhangmin - thanks for the reviews!

> Can you please share the stacktrace of the error?

@eolivelli - not sure I understand this? The code wasn't throwing bugs before the fix, it was handling exceptions incorrectly, so I am not sure which stacktrace are you referring to. LMK if I missed something!

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java [hunk 3]
```diff
@@ -106,6 +106,7 @@ private void checkForPartitionsChanges(Timeout timeout) {
         start().whenComplete((tw, ex) -> {
            if (ex != null) {
                log.warn("Failed to check for changes in number of partitions");
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:108]: I think @eolivelli was meaning to include the exception in the log message here.

**Review by merlimat — APPROVED** [review]

**Comment by ZiyaoWei** [comment]
[](https://github.com/merlimat)/pulsarbot run-failure-checks

**Review by eolivelli — APPROVED** [review]

**Review by gaozhangmin — DISMISSED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java [hunk 4]
```diff
@@ -236,27 +237,27 @@ private void readAllExistingMessages(Reader<T> reader, CompletableFuture<Reader<
...
-                       long endTime = System.nanoTime();
-                       long durationMillis = TimeUnit.NANOSECONDS.toMillis(endTime - startTime);
-                       log.info("Started table view for topic {} - Replayed {} messages in {} seconds",
-                               reader.getTopic(),
-                               messagesRead,
-                               durationMillis / 1000.0);
-                       future.complete(reader);
-                       readTailMessages(reader);
-                   }
+                    if (hasMessage) {
```
Comment (Demogorgon314) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:240]: Please undo the unrelated change.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TableViewTest.java [hunk 5]
```diff
@@ -158,4 +162,57 @@ public void testTableViewUpdatePartitions() throws Exception {
         });
         Assert.assertEquals(tv.keySet(), keys2);
     }
+
+
+    @Test(timeOut = 30 * 1000)
+    // Regression test for making sure partition changes are always periodically checked even after a check returned
+    // exceptionally.
+    public void testTableViewUpdatePartitionsTriggeredDespiteExceptions() throws Exception {
+        String topic = "persistent://public/default/tableview-test-update-partitions";
```
Comment (Demogorgon314) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TableViewTest.java:171]: We should use another topic name here, this topic name is used in the previous test method.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TableViewTest.java [hunk 6]
```diff
@@ -86,7 +90,7 @@ protected void cleanup() throws Exception {
         try (Producer<byte[]> producer = builder.create()) {
             CompletableFuture<?> lastFuture = null;
             for (int i = 0; i < count; i++) {
-                String key = "key"+ i;
+                String key = "key" + i;
```
Comment (Demogorgon314) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TableViewTest.java:93]: Please undo the unrelated change.

Comment (ZiyaoWei) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TableViewTest.java:93]: Uh, my autoformatter somehow kicked in again, thanks for catching!

**Review by merlimat — APPROVED** [review]
